services/src/main/java/org/keycloak/protocol/saml/SamlMetadataPublicKeyLoader.java

```diff
@@ -0,0 +1,142 @@
+import javax.xml.crypto.dsig.keyinfo.KeyInfo;
+import javax.xml.crypto.dsig.keyinfo.KeyName;
+import javax.xml.crypto.dsig.keyinfo.X509Data;
+import org.jboss.logging.Logger;
+import org.keycloak.connections.httpclient.HttpClientProvider;
+import org.keycloak.crypto.KeyUse;
+import org.keycloak.crypto.KeyWrapper;
+import org.keycloak.crypto.PublicKeysWrapper;
+import org.keycloak.dom.saml.v2.metadata.EntityDescriptorType;
+import org.keycloak.dom.saml.v2.metadata.IDPSSODescriptorType;
+import org.keycloak.dom.saml.v2.metadata.KeyDescriptorType;
+import org.keycloak.dom.saml.v2.metadata.KeyTypes;
+import org.keycloak.dom.saml.v2.metadata.SPSSODescriptorType;
+import org.keycloak.keys.PublicKeyLoader;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.saml.processing.core.saml.v2.util.SAMLMetadataUtil;
+import org.keycloak.saml.processing.core.util.XMLSignatureUtil;
+import org.w3c.dom.Element;
+
+/**
+ * <p>PublicKeyLoader to retrieve keys from a SAML metadata entity endpoint.
+ * It can be used to load IDP or SP keys.</p>
+ *
+ * @author rmartinc
+ */
+public class SamlMetadataPublicKeyLoader implements PublicKeyLoader {
+
+    private static final Logger logger = Logger.getLogger(SamlMetadataPublicKeyLoader.class);
+    private final KeycloakSession session;
+    private final String metadataUrl;
+    private final boolean forIdP;
+
+    public SamlMetadataPublicKeyLoader(KeycloakSession session, String metadataUrl) {
+        this(session, metadataUrl, true);
+    }
+
+    public SamlMetadataPublicKeyLoader(KeycloakSession session, String metadataUrl, boolean forIdP) {
+        this.session = session;
+        this.metadataUrl = metadataUrl;
+        this.forIdP = forIdP;
+    }
+
+    @Override
+    public PublicKeysWrapper loadKeys() throws Exception {
+        logger.debugf("loading keys from metadata endpoint %s", metadataUrl);
+        InputStream inputStream = session.getProvider(HttpClientProvider.class).get(metadataUrl);
+
+        List<KeyDescriptorType> keyDescriptor;
+        EntityDescriptorType entityType = SAMLMetadataUtil.parseEntityDescriptorType(inputStream);
+        if (forIdP) {
+            IDPSSODescriptorType idpDescriptor = SAMLMetadataUtil.locateIDPSSODescriptorType(entityType);
+            keyDescriptor = idpDescriptor != null? idpDescriptor.getKeyDescriptor() : null;
+        } else {
+            SPSSODescriptorType spDescriptor = SAMLMetadataUtil.locateSPSSODescriptorType(entityType);
+            keyDescriptor = spDescriptor != null? spDescriptor.getKeyDescriptor() : null;
+        }
+
+        List<KeyWrapper> keys = new ArrayList<>();
+        if (keyDescriptor != null) {
+            for (KeyDescriptorType keyDescriptorType : keyDescriptor) {
+                Element keyInfoElement = keyDescriptorType.getKeyInfo();
+                if (keyInfoElement == null) {
+                    continue;
+                }
+
+                KeyUse use = KeyUse.SIG; // TODO: default SIG? Or Both?
```

Comment:
Perhaps no default and if use is not set it means it can be used for both. That makes more sense to me than be opinionated about the usage of a key.

Reply:
Done! No, it was not prepared but it's now. If teh use is null the keys is returned for both uses.